File: frontend/src/services/rt_structure.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { api } from './api';
import { getTenantSlugOrFallback } from '../utils/tenant';

export interface RTPeriod {
  id: string;
  tenant_id: string;
  name: string;
  start_date: string;
  end_date: string;
  status: 'draft' | 'active' | 'archived';
  sk_number?: string;
  sk_file_url?: string;
  created_at: string;
  updated_at: string;
}

export interface RTMember {
  id: string;
  period_id: string;
  resident_id: string;
  role: 'ketua' | 'wakil' | 'sekretaris' | 'bendahara' | 'seksi' | 'penasihat' | string;
  section?: string;
  custom_title?: string;
  phone_override?: string;
  photo_url?: string;
  status: 'aktif' | 'demisioner' | 'nonaktif';
  joined_at: string;
  created_at: string;
  updated_at: string;
  resident_name?: string;
  resident_nik?: string;
  phone?: string;
}

// Queries
export function useRTPeriods(status?: string) {
  return useQuery<{ data: RTPeriod[] }>({
    queryKey: ['rt-structure', 'periods', status],
    queryFn: async () => {
      const res = await api.get<{ data: RTPeriod[] }>('/rt-structure/periods', {
        params: status ? { status } : undefined,
      });
      return res.data;
    },
  });
}

export function useRTMembers(periodId: string, filters?: { section?: string; role?: string; status?: string }) {
  return useQuery<{ data: RTMember[] }>({
    queryKey: ['rt-structure', 'members', periodId, filters],
    queryFn: async () => {
      if (!periodId) return { data: [] };
      const res = await api.get<{ data: RTMember[] }>('/rt-structure/members', {
        params: { period_id: periodId, ...filters },
      });
      return res.data;
    },
    enabled: Boolean(periodId),
  });
}

export function usePublicRTStructureQuery() {
  const slug = getTenantSlugOrFallback();
  return useQuery<{ period: RTPeriod; members: RTMember[] }>({
    queryKey: ['public-rt-structure', slug],
    queryFn: async () => {
      const res = await api.get<{ period: RTPeriod; members: RTMember[] }>(`/t/${slug}/rt-structure`);
      return res.data;
    },
    enabled: Boolean(slug),
  });
}

// Mutations
export function useCreateRTPeriod() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: {
      name: string;
      start_date: string;
      end_date: string;
      status: string;
      sk_number?: string;
      sk_file_url?: string;
    }) => {
      const res = await api.post<RTPeriod>('/rt-structure/periods', payload);
      return res.data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['rt-structure'] });
    },
  });
}

export function useUpdateRTPeriod() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({
      id,
      ...payload
    }: {
      id: string;
      name: string;
      start_date: string;
      end_date: string;
      status: string;
      sk_number?: string;
      sk_file_url?: string;
    }) => {
      const res = await api.put<RTPeriod>(`/rt-structure/periods/${id}`, payload);
      return res.data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['rt-structure'] });
    },
  });
}

export function useAddRTMember() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: {
      period_id: string;
      resident_id: string;
      role: string;
      section?: string;
      custom_title?: string;
      phone_override?: string;
      photo_url?: string;
      status: string;
    }) => {
      const res = await api.post<RTMember>('/rt-structure/members', payload);
      return res.data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['rt-structure'] });
    },
  });
}

export function useUpdateRTMember() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({
      id,
      ...payload
    }: {
      id: string;
      role: string;
      section?: string;
      custom_title?: string;
      phone_override?: string;
      photo_url?: string;
      status: string;
    }) => {
      const res = await api.put<RTMember>(`/rt-structure/members/${id}`, payload);
      return res.data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['rt-structure'] });
    },
  });
}

export function useDeleteRTMember() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      await api.delete(`/rt-structure/members/${id}`);
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['rt-structure'] });
    },
  });
}

File: backend/internal/domain/rt_structure.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type RTPeriod struct {
	ID        uuid.UUID  `json:"id"`
	TenantID  uuid.UUID  `json:"tenant_id"`
	Name      string     `json:"name"`
	StartDate time.Time  `json:"start_date"`
	EndDate   time.Time  `json:"end_date"`
	Status    string     `json:"status"` // draft, active, archived
	SKNumber  *string    `json:"sk_number,omitempty"`
	SKFileURL *string    `json:"sk_file_url,omitempty"`
	CreatedBy *uuid.UUID `json:"created_by,omitempty"`
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`

	Members []*RTMember `json:"members,omitempty"`
}

type RTMember struct {
	ID            uuid.UUID `json:"id"`
	PeriodID      uuid.UUID `json:"period_id"`
	ResidentID    uuid.UUID `json:"resident_id"`
	Role          string    `json:"role"` // ketua, wakil, sekretaris, bendahara, seksi, penasihat
	Section       *string   `json:"section,omitempty"`
	CustomTitle   *string   `json:"custom_title,omitempty"`
	PhoneOverride *string   `json:"phone_override,omitempty"`
	PhotoURL      *string   `json:"photo_url,omitempty"`
	Status        string    `json:"status"` // aktif, demisioner, nonaktif
	JoinedAt      time.Time `json:"joined_at"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`

	// Relasi resident
	ResidentName string  `json:"resident_name,omitempty"`
	ResidentNIK  string  `json:"resident_nik,omitempty"`
	Phone        *string `json:"phone,omitempty"`
}

type RTStructureRepository interface {
	// Periode
	CreatePeriod(ctx context.Context, p *RTPeriod) error
	GetPeriodByID(ctx context.Context, tenantID, id uuid.UUID) (*RTPeriod, error)
	GetActivePeriod(ctx context.Context, tenantID uuid.UUID) (*RTPeriod, error)
	ListPeriods(ctx context.Context, tenantID uuid.UUID, status string) ([]*RTPeriod, error)
	UpdatePeriod(ctx context.Context, p *RTPeriod) error

	// Members
	AddMember(ctx context.Context, m *RTMember) error
	GetMemberByID(ctx context.Context, tenantID, id uuid.UUID) (*RTMember, error)
	ListMembers(ctx context.Context, tenantID, periodID uuid.UUID, section, role, status string) ([]*RTMember, error)
	UpdateMember(ctx context.Context, m *RTMember) error
	DeleteMember(ctx context.Context, tenantID, id uuid.UUID) error
}

type RTStructureUsecase interface {
	CreatePeriod(ctx context.Context, tenantID uuid.UUID, p *RTPeriod) error
	GetPeriod(ctx context.Context, tenantID, id uuid.UUID) (*RTPeriod, error)
	GetActivePeriod(ctx context.Context, tenantID uuid.UUID) (*RTPeriod, error)
	ListPeriods(ctx context.Context, tenantID uuid.UUID, status string) ([]*RTPeriod, error)
	UpdatePeriod(ctx context.Context, tenantID uuid.UUID, p *RTPeriod) error

	AddMember(ctx context.Context, tenantID uuid.UUID, m *RTMember) error
	ListMembers(ctx context.Context, tenantID, periodID uuid.UUID, section, role, status string) ([]*RTMember, error)
	UpdateMember(ctx context.Context, tenantID uuid.UUID, m *RTMember) error
	DeleteMember(ctx context.Context, tenantID, id uuid.UUID) error
}
